SwrveSDK/src/main/java/com/swrve/sdk/SwrveImp.java

```diff
@@ -970,36 +976,36 @@ private boolean supportsDeviceFilter(String requirement) {
     }
 
     protected void downloadAssets(final Set<String> assetsQueue) {
-        assetsCurrentlyDownloading = true;
+        assetManager.setAssetsCurrentlyDownloading(true);
         ExecutorService resourceDownloadExecutor = Executors.newSingleThreadExecutor();
         resourceDownloadExecutor.execute(SwrveRunnables.withoutExceptions(new Runnable() {
             @Override
             public void run() {
                 Set<String> assetsToDownload = filterExistingFiles(assetsQueue);
                 for (String asset : assetsToDownload) {
+
                     boolean success = downloadAssetSynchronously(asset);
                     if (success) {
-                        synchronized (assetsOnDisk) {
-                            assetsOnDisk.add(asset);
+                        if (success) {
+                            assetManager.addAsset(asset);
+                        } else {
+                            assetManager.addFailedAsset(asset);
                         }
                     }
+
+                    if (assetManager.hasFailedAssets()) {
+                        downloadAssets(assetManager.getFailedAssets());
+                    }
                 }
-                assetsCurrentlyDownloading = false;
+                assetManager.setAssetsCurrentlyDownloading(false);
                 autoShowMessages();
             }
         }));
     }
 
-    protected SwrveCampaign loadCampaignFromJSON(JSONObject campaignData, Set<String> assetsQueue) throws JSONException {
-        return new SwrveCampaign((SwrveBase<?, ?>) this, campaignData, assetsQueue);
-    }
-
-    protected SwrveConversationCampaign loadConversationCampaignFromJSON(JSONObject campaignData, Set<String> assetsQueue) throws JSONException {
-        return new SwrveConversationCampaign((SwrveBase<?, ?>) this, campaignData, assetsQueue);
-    }
-
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
     protected boolean downloadAssetSynchronously(final String assetPath) {
```

Comment:
We should move the downloadAssetSynchronously to the SwrveAssetManager also and rename it just downloadAsset.